### ISSUE TITLE
[HOLD for payment 2024-06-28] [$250] [CRITICAL] [UX Reliability] The chat finder page shows duplicated result for a moment 

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#43094]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:**
**Reproducible in staging?:** Y
**Reproducible in production?:** Y 
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C05LX9D6E07/p1717508731782149

## Action Performed:
Break down in numbered steps

1. Create Workspace as Alice
1. Add couple members including Bob
1. Create couple rooms on that workspace
1. Sign into an account off Bob
1. Turn on the focus mode
1. Sign out and sign back in as bob
1. In chat search page search for the name of the room on the policy
1. Open it the room from the Search page

## Expected Result:
Describe what you think should've happened

There should only be one row for that chat

## Actual Result:
Describe what actually happened

The specific chat is duplicated in the search results

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~0157768929ce5b2314</li>
        <li>Upwork Job ID: 1798292023932084224</li>
        <li>Last Price Increase: 2024-06-05</li>
<li>Automatic offers: </li>
<ul>
<li>ahmedGaber93 | Reviewer | 102621261</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @dylanexpensify</details>

**Comment by mountiny** [comment]
Nice @gedu can you check out the proposal if it matches your findings?

**Comment by ahmedGaber93** [comment]
@tienifr your solution is works, but I think your RCA is not correct, log `missingReportIds` before your fix seem correct, and after applied your fix it always logs empty array. Maybe we need to find the correct root cause first.

**Comment by ahmedGaber93** [comment]
This proposal from 2 min ago https://github.com/Expensify/App/issues/43049#issuecomment-2149700859 in the other issue add RTC about what I am thinking.

**Comment by gedu** [comment]
@mountiny 
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
After creating a room and searching for it, the room is displayed twice in the search results (no cache).

### What is the root cause of that problem?
There are two `useEffect` hooks listening to the same object `reports`. The first hook updates the `options` with the new list of options inside `setOptions` using the state provided by the function. The second hook uses prevReports to find a missing reportID (which in this case is the new report), and updates the same `setOptions` state, adding the "missing" report to the already updated options (added by the first useEffect), cause the duplication.

### What changes do you think we should make in order to solve the problem?
Since the first `useEffect` always recreates the options when `reports` change, the second `useEffect` detects a difference and updates the state over an already updated one. To solve this, unify the two `useEffect` hooks to update the report `options` whenever the `reports` object changes, avoiding extra logic and potential re-renders.

### What alternative solutions did you explore? (Optional)
Before pushing the missing report ID found by the filter, in the second useEffect, check if the reportID is already in the option list.

I tested some older issues, like the https://github.com/Expensify/App/issues/39705 and the main fix doesn't cause a regression on that side, I'm not sure if there is another bug to address

**Comment by ahmedGaber93** [comment]
I think this issue is a regression from https://github.com/Expensify/App/pull/41636 which delete the return early on the first `useEffect`, and this make the new reports are handling in two `useEffect`
```diff
- const lastUpdatedReport = ReportUtils.getLastUpdatedReport();

- if (!lastUpdatedReport) {
-       return;
- }
```

> To solve this, unify the two useEffect hooks to update the report options whenever the reports object changes

@gedu I agree, we need to make sure the new report is handled once, but not necessarily we need to unify the two useEffect, we can also keep them separator and prevent interference between them. @gedu Can you share how you do that?

> Before pushing the missing report ID found by the filter, in the second useEffect, check if the reportID is already in the option list.

I think this will always cancel the second `useEffect` work, as the first `useEffect` always add it first.

**Comment by ahmedGaber93** [comment]
Yeah, but `missingReportIDs` will be always empty array as the first `useEffect` add them first, so it seems we completely remove the second `useEffect`

**Comment by ahmedGaber93** [comment]
The RCA here is we have two `useEffect` will add the new report to the list:

 -  [the second useEffect](https://github.com/Expensify/App/blob/47d5f5f049b8eb0be117e4fd91ab0fee9b834134/src/components/OptionListContextProvider.tsx#L77-L102) job is handling the new items and it works fine

 - [the first one](https://github.com/Expensify/App/blob/47d5f5f049b8eb0be117e4fd91ab0fee9b834134/src/components/OptionListContextProvider.tsx#L58-L75) job is updating an existing item when a report is updated. And it shouldn't handle adding new items, but PR https://github.com/Expensify/App/pull/41636 removed the return early when no existing item need to update, and this let it adding the new items and cause this regression



@dylanexpensify As no proposal is accepted yet, let's assign @gedu from callstack.

🎀 👀 🎀 C+ reviewed

**Comment by gedu** [comment]
> @gedu I agree, we need to make sure the new report is handled once, but not necessarily we need to unify the two useEffect, we can also keep them separator and prevent interference between them. @gedu Can you share how you do that?

I created a draft PR: https://github.com/Expensify/App/pull/43164, so you can check the change.


https://github.com/Expensify/App/assets/1676818/514aa2d8-872e-42fa-b97d-32a80b351ff8

**Comment by ahmedGaber93** [comment]
@gedu Thanks for the updates.

The fix you propose seem as removing the second `useEffect`, but I think will lead to reduce performance as this will lead to `createOptionList` for the whole reports list every time an item removed or edited.

Can we find a solution for this and keep the current two `useEffect`?
 - The first one will `createOptionFromReport` only for the updated report.
 - the second will `createOptionFromReport` once for the whole reports in the first time, and then when item added/removed, it will `createOptionFromReport` only for it.
 
 I mean fix it as regression if https://github.com/Expensify/App/pull/41636

**Comment by gedu** [comment]
@ahmedGaber93 but the `createOptionList` is always getting call, everytime the `reports` changed both useEffect get executed, and the first one always run the `createOptionList` 
I just tested, doing some simple search (cached and no cache values), creating a task and searching, creating a workspace assign a task and then delete it

```
[CHATFINDERPAGE-CONTEXT] 1) <createOptionList>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <missingReportIds>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 1) <createOptionList>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <missingReportIds>
2index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <createOptionFromReport>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 1) <createOptionList>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <missingReportIds>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <createOptionFromReport>
index.ts:77 Flushing Onyx updates queue (15) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
index.ts:77 [CHATFINDERPAGE-CONTEXT] 1) <createOptionList>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <missingReportIds>
[CHATFINDERPAGE] filteredOptions {recentReports: Array(0), personalDetails: Array(0), userToInvite: null, headerMessage: 'No results found'}
index.ts:77 [CHATFINDERPAGE-CONTEXT] 1) <createOptionList>
index.ts:77 [CHATFINDERPAGE-CONTEXT] 2) <missingReportIds>
```
Not sure what performance impact my fix might cause, given that `createOptionList` is called all the time.
Can you point me out to a case where the first useEffect isn't being called and the second useEffect yes

Record a video for the regression you said and my fix doesn't cause a regression
https://github.com/Expensify/App/assets/1676818/cf67de3c-0f77-4ccc-a6c5-46f079b85ce5

**Comment by ahmedGaber93** [comment]
> but the createOptionList is always getting call

@gedu Yes, but this is part of the regression by #41636 that we need to fix it here

This regression remove the return early in the first useEffect and this caused the current issue and also the performance issue, so I think after fix it, the both issues should be fixed.

**Comment by gedu** [comment]
Yes, I will take a look, sorry for the ping pong question, trying to wrap my head on this because with just one `useEffect` fix all the issues, will dig deeper

**Comment by gedu** [comment]
@ahmedGaber93 I created a new function to listen when a report changes, I listen to them individually so I can save them into a Set and when I use the updated IDs I clear the Set. That way when a report changes I can know which one change, I delete it from the prevReports, and the second `useEffect` handle the missingReportID. 
The issue that I'm having is, when an update changes in this case changing the name of the Workspace, it seems that all the rooms depending on it aren't sent via Onyx, same issue as above, but in this case the room changed but not the Workspace.
I'm not finding a predictable way to know which report changes so I can remove it from the list and later create the optionItem just for that items.

<img width="363" alt="Screenshot 2024-06-11 at 14 31 04" src="https://github.com/Expensify/App/assets/1676818/d23e1663-8a26-4ded-9b90-f6ad93f72072">

I would start by considering my proposal to unify and update the list continuously. Additionally, I will work on a follow-up PR to improve the performance of the `createOptionList`, given that the only way to know when a report changes is through Onyx, and I can't get reliable values.

**Comment by gedu** [comment]
@ahmedGaber93 I make my PR ready for review https://github.com/Expensify/App/pull/43164

**Comment by dylanexpensify** [comment]
waiting for regression period

**Comment by ahmedGaber93** [comment]
**BugZero Checklist**: The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [x] [@ahmedGaber93] The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/41636
- [x] [@ahmedGaber93] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/41636/files#r1660214150
- [x] [@ahmedGaber93] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion: N/A
- [x] [@ahmedGaber93] Determine if we should create a regression test for this bug.
- [x] [@ahmedGaber93] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

### Regression Test Proposal

1. Open the App
2. Click Account Settings > Workspaces > and create new workspace
3. Click fab icon > Start chat > Room tab
4. Fill room name and select the workspace you create it in step 1 then click create room
5. Open Search page and Verify that the room you created in step 4 appear once in search result

Do we agree 👍 or 👎

**Comment by dylanexpensify** [comment]
Payment summary:
- contributor+ @ahmedGaber93 $250

Please apply, request!

**Comment by dylanexpensify** [comment]
Done!
